File: rag-langchain-python/word_query_retriever.py
from typing import List
from langchain_core.documents import Document
from langchain_core.retrievers import BaseRetriever
from marklogic import Client


class WordQueryRetriever(BaseRetriever):
    client: Client

    @classmethod
    def create(cls, client: Client):
        return cls(client=client)

    def _get_relevant_documents(self, query: str) -> List[Document]:
        words = []
        for word in query.split():
            if len(word) > 2:
                words.append(word.lower().replace("?", ""))

        word_query = "<word-query xmlns='http://marklogic.com/cts'>"
        for word in words:
            word_query = f"{word_query}<text>{word}</text>"
        word_query = f"{word_query}</word-query>"

        print(f"Searching with query: {word_query}")
        results = self.client.documents.search(
            query=word_query,
            page_length=10,
            collections=["events"],
        )
        print(f"Count of MarkLogic documents sent to the LLM: {len(results)}")
        for result in results:
            print(f"URI: {result.uri}")
        return map(lambda doc: Document(page_content=doc.content["transcript"]), results)

File: rag-langchain-js/wordQueryRetriever.js
import { BaseRetriever } from "@langchain/core/retrievers";
import { Document } from "@langchain/core/documents";
import pkg from 'marklogic';
const { ctsQueryBuilder, queryBuilder } = pkg;

export class WordQueryRetriever extends BaseRetriever {
  lc_namespace = ["langchain", "retrievers"];
  marklogicClient;

  constructor(fields) {
    super(fields);
    this.marklogicClient = fields.marklogicClient;
  }

  #buildWordsArray(chat_question) {
    const words = [];
    chat_question.split(" ").forEach(word => {
      if (word.length > 2) words.push(word.toLowerCase().replace("?", ""))
    })
    return words;
  }

  async _getRelevantDocuments(
    chat_question, _callbacks,
  ) {
    const words = this.#buildWordsArray(chat_question);
    const wordQuery = queryBuilder.where(ctsQueryBuilder.cts.wordQuery(words));
    const results = await this.marklogicClient.documents.query(wordQuery).result((queryResults) => {
      return queryResults;
    });

    const documents = [];
    results.forEach((result) => {
      console.log(result["uri"])
      documents.push(
        new Document({
          pageContent: result['content']['transcript'],
          metadata: {"uri": result["uri"]}
        })
      )
    });
    return documents;
  }
}
